cluster-autoscaler/processors/scaledowncandidates/scale_down_candidates_delay_processor.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scaledowncandidates
+
+import (
+	"reflect"
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+
+	"k8s.io/autoscaler/cluster-autoscaler/clusterstate"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+)
+
+// ScaleDownCandidatesDelayProcessor is a processor to filter out
+// nodes according to scale down delay per nodegroup
+type ScaleDownCandidatesDelayProcessor struct {
+	clusterStateRegistry *clusterstate.ClusterStateRegistry
+}
+
+// GetPodDestinationCandidates returns nodes as is no processing is required here
+func (p *ScaleDownCandidatesDelayProcessor) GetPodDestinationCandidates(ctx *context.AutoscalingContext,
+	nodes []*apiv1.Node) ([]*apiv1.Node, errors.AutoscalerError) {
+	return nodes, nil
+}
+
+// GetScaleDownCandidates returns filter nodes based on if scale down is enabled or disabled per nodegroup.
+func (p *ScaleDownCandidatesDelayProcessor) GetScaleDownCandidates(ctx *context.AutoscalingContext,
+	nodes []*apiv1.Node) ([]*apiv1.Node, errors.AutoscalerError) {
+	result := []*apiv1.Node{}
+
+	for _, node := range nodes {
+		nodeGroup, err := ctx.CloudProvider.NodeGroupForNode(node)
+		if err != nil {
+			klog.Warningf("Error while checking node group for %s: %v", node.Name, err)
+			continue
+		}
+		if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
+			klog.V(4).Infof("Node %s should not be processed by cluster autoscaler (no node group config)", node.Name)
+			continue
+		}
+
+		lastScaledUps := p.clusterStateRegistry.GetScaleUpRequests()
+		lastScaleDowns := p.clusterStateRegistry.GetScaleDownRequests()
+		lastScaleUpFailures := p.clusterStateRegistry.GetScaleUpFailures()
+		currentTime := time.Now()
+
+		if lastScaledUps[nodeGroup.Id()] != nil &&
+			lastScaledUps[nodeGroup.Id()].ExpectedAddTime.After(currentTime) {
+			klog.V(4).Infof("Skipping scale down on node group %s because it was scaled up recently at %v", nodeGroup.Id(), lastScaledUps[nodeGroup.Id()].Time)
```

Comment:
I have set the log level at v4 but that might not be the best thing to do. It might help to have this at v2 so that the users know what is happening around scale down. Any feedback on this is welcome :pray: